File: api/src/schemas/resource_pool.py
"""Pydantic schemas for resource pools and cross-program availability."""

from __future__ import annotations

from datetime import date, datetime  # noqa: TC003 - Required at runtime for Pydantic
from decimal import Decimal
from typing import Annotated, Any
from uuid import UUID  # noqa: TC003 - Required at runtime for Pydantic

from pydantic import BaseModel, ConfigDict, Field

from src.models.resource_pool import PoolAccessLevel

# =============================================================================
# Resource Pool Schemas
# =============================================================================


class ResourcePoolCreate(BaseModel):
    """Schema for creating a resource pool."""

    name: Annotated[str, Field(min_length=1, max_length=100, description="Pool name")]
    code: Annotated[
        str,
        Field(
            min_length=1,
            max_length=50,
            pattern=r"^[A-Z0-9\-_]+$",
            description="Unique pool code",
        ),
    ]
    description: str | None = Field(default=None, description="Pool description")


class ResourcePoolUpdate(BaseModel):
    """Schema for updating a resource pool."""

    name: str | None = Field(default=None, min_length=1, max_length=100)
    description: str | None = None
    is_active: bool | None = None


class ResourcePoolResponse(BaseModel):
    """Response schema for a resource pool."""

    id: UUID
    name: str
    code: str
    description: str | None
    owner_id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =============================================================================
# Pool Member Schemas
# =============================================================================


class PoolMemberCreate(BaseModel):
    """Schema for adding a resource to a pool."""

    resource_id: UUID = Field(..., description="Resource to add to pool")
    allocation_percentage: Annotated[
        Decimal,
        Field(
            default=Decimal("100.00"),
            ge=Decimal("0"),
            le=Decimal("100"),
            description="Percentage of resource available to pool (0-100)",
        ),
    ] = Decimal("100.00")


class PoolMemberResponse(BaseModel):
    """Response schema for a pool member."""

    id: UUID
    pool_id: UUID
    resource_id: UUID
    allocation_percentage: Decimal
    is_active: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =============================================================================
# Pool Access Schemas
# =============================================================================


class PoolAccessCreate(BaseModel):
    """Schema for granting pool access to a program."""

    program_id: UUID = Field(..., description="Program to grant access")
    access_level: PoolAccessLevel = Field(
        default=PoolAccessLevel.VIEWER,
        description="Access level for the program",
    )


class PoolAccessResponse(BaseModel):
    """Response schema for pool access grant."""

    id: UUID
    pool_id: UUID
    program_id: UUID
    access_level: PoolAccessLevel
    granted_by: UUID | None
    granted_at: datetime

    model_config = ConfigDict(from_attributes=True)


# =============================================================================
# Availability Schemas
# =============================================================================


class PoolAvailabilityResponse(BaseModel):
    """Response schema for pool availability check."""

    pool_id: UUID
    pool_name: str
    date_range_start: date
    date_range_end: date
    resources: list[dict[str, Any]]
    conflict_count: int
    conflicts: list[dict[str, Any]]


class ConflictCheckRequest(BaseModel):
    """Request schema for checking assignment conflicts."""

    resource_id: UUID = Field(..., description="Resource to check")
    program_id: UUID = Field(..., description="Program making the assignment")
    start_date: date = Field(..., description="Assignment start date")
    end_date: date = Field(..., description="Assignment end date")
    units: Annotated[
        Decimal,
        Field(
            default=Decimal("1.00"),
            ge=Decimal("0"),
            le=Decimal("10"),
            description="Allocation units",
        ),
    ] = Decimal("1.00")


class ConflictCheckResponse(BaseModel):
    """Response schema for conflict check."""

    has_conflicts: bool
    conflict_count: int
    conflicts: list[dict[str, Any]]

File: web/src/types/resourcePool.ts
export enum PoolAccessLevel {
  VIEWER = "VIEWER",
  EDITOR = "EDITOR",
  MANAGER = "MANAGER",
}

export interface ResourcePoolCreate {
  name: string;
  code: string;
  description?: string;
}

export interface ResourcePoolUpdate {
  name?: string;
  description?: string;
  is_active?: boolean;
}

export interface ResourcePoolResponse {
  id: string;
  name: string;
  code: string;
  description: string | null;
  owner_id: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface PoolMemberCreate {
  resource_id: string;
  allocation_percentage?: number;
}

export interface PoolMemberResponse {
  id: string;
  pool_id: string;
  resource_id: string;
  allocation_percentage: string;
  is_active: boolean;
  created_at: string;
}

export interface PoolAccessCreate {
  program_id: string;
  access_level: PoolAccessLevel;
}

export interface PoolAccessResponse {
  id: string;
  pool_id: string;
  program_id: string;
  access_level: PoolAccessLevel;
  granted_by: string | null;
  granted_at: string;
}

export interface PoolAvailabilityResource {
  resource_id: string;
  resource_name: string;
  available_hours: number;
  assigned_hours: number;
}

export interface PoolConflict {
  resource_id: string;
  resource_name: string;
  conflict_date: string;
  programs_involved: string[];
  overallocation_hours: number;
}

export interface PoolAvailabilityResponse {
  pool_id: string;
  pool_name: string;
  date_range_start: string;
  date_range_end: string;
  resources: PoolAvailabilityResource[];
  conflict_count: number;
  conflicts: PoolConflict[];
}

export interface ConflictCheckRequest {
  resource_id: string;
  program_id: string;
  start_date: string;
  end_date: string;
  units?: number;
}

export interface ConflictCheckResponse {
  has_conflicts: boolean;
  conflict_count: number;
  conflicts: Array<{
    date: string;
    programs: string[];
    overallocation_hours: number;
  }>;
}
